File: go/module-config/push-to-function/cmd/run-foo-bar-override/run-foo-bar-override.go
/*
 *  Copyright © 2021 Paulo Villela. All rights reserved.
 *  Use of this source code is governed by the Apache 2.0 license
 *  that can be found in the LICENSE file.
 */

package main

import (
	"github.com/pvillela/module-config/go/arch/util"
	"github.com/pvillela/module-config/go/module-config/push-to-function/config"
	"github.com/pvillela/module-config/go/module-config/push-to-function/fs"
	"github.com/pvillela/module-config/go/module-config/push-to-function/fs/boot"
)

func main() {
	boot.FooSflCfgAdapter = func(src config.AppCfgSrc) func() fs.FooSflCfgInfo {
		return util.ThunkOf(fs.FooSflCfgInfo{X: "foo"})
	}

	boot.BarBfCfgAdapter = func(src config.AppCfgSrc) func() fs.BarBfCfgInfo {
		return util.ThunkOf(fs.BarBfCfgInfo{Z: 99})
	}

	{
		fooSfl := boot.FooSflBoot(nil)
		fooSfl()
	}

	{
		fooSfl := boot.FooSflBoot(config.GetAppConfiguration)
		fooSfl()
	}
}

File: kt/main/kotlin/moduleconfig/pushtofunction/run/RunFooBarOverride.kt
/*
 *  Copyright © 2021 Paulo Villela. All rights reserved.
 *  Use of this source code is governed by the Apache 2.0 license
 *  that can be found in the LICENSE file.
 */

package tryout.moduleconfig.pushtofunction.run

import tryout.moduleconfig.pushtofunction.config.getAppConfiguration
import tryout.moduleconfig.pushtofunction.fs.BarBfCfgInfo
import tryout.moduleconfig.pushtofunction.fs.FooSflCfgInfo
import tryout.moduleconfig.pushtofunction.fs.boot.barBfCfgAdapter
import tryout.moduleconfig.pushtofunction.fs.boot.fooSflBoot
import tryout.moduleconfig.pushtofunction.fs.boot.fooSflCfgAdapter

fun main() {
	fooSflCfgAdapter = { appSrc -> { FooSflCfgInfo(x = "foo") } }

	barBfCfgAdapter =  { appSrc -> { BarBfCfgInfo(z = 99) } }

	run {
		val fooSfl = fooSflBoot(null)
		fooSfl()
	}

	run {
		val fooSfl = fooSflBoot(::getAppConfiguration)
		fooSfl()
	}
}
